From e36e38fa13225695af438e73702a4f5447b8c6e2 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 27 Apr 2026 09:42:20 +0800
Subject: [PATCH] mtk: hostapd: detect missing for ATTLM input parameters

To trigger an ATTLM, the user must provide the switch time, duration,
and disabled links. Detect if any of these parameters are missing.

Also, set the default value of link_mapping_size to 0, so that
link_mapping_size can be omitted.

fixup! mtk: hostapd: add A-TTLM support

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 hostapd/ctrl_iface.c | 24 ++++++++++++++++++++----
 1 file changed, 20 insertions(+), 4 deletions(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 2fd056b41..4501edc06 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -4007,9 +4007,10 @@ static int hostapd_ctrl_iface_set_attlm(struct hostapd_data *hapd, char *cmd,
 	struct attlm_settings *attlm;
 	struct hostapd_data *h;
 	char *token, *context = NULL;
-	u16 switch_time, disabled_links, valid_links;
-	u32 duration;
-	int ret, i, link_mapping_size;
+	u16 switch_time = MAX_SWITCH_TIME_MS + 1;
+	u16 disabled_links = BIT(MAX_NUM_MLD_LINKS) + 1, valid_links;
+	u32 duration = MAX_DURATION_MS + 1;
+	int ret, i, link_mapping_size = 0;
 
 	if (!hapd->conf->mld_ap || !hapd->mld)
 		return -1;
@@ -4072,11 +4073,26 @@ static int hostapd_ctrl_iface_set_attlm(struct hostapd_data *hapd, char *cmd,
 				continue;
 		}
 
-		wpa_printf(MSG_INFO, "CTRL: Invalid SET_ATTLM parameter: %s",
+		wpa_printf(MSG_ERROR, "CTRL: Invalid SET_ATTLM parameter: %s",
 			   token);
 		return -1;
 	}
 
+	if (switch_time == MAX_SWITCH_TIME_MS + 1) {
+		wpa_printf(MSG_ERROR, "CTRL: switch time is not set");
+		return -1;
+	}
+
+	if (duration == MAX_DURATION_MS + 1) {
+		wpa_printf(MSG_ERROR, "CTRL: duration is not set");
+		return -1;
+	}
+
+	if (disabled_links == BIT(MAX_NUM_MLD_LINKS) + 1) {
+		wpa_printf(MSG_ERROR, "CTRL: disabled links is not set");
+		return -1;
+	}
+
 	wpa_printf(MSG_DEBUG,
 		   "MLD: Adv-TTLM request disabled_links=%u, link_mapping_size=%d, switch_time=%u, duration=%u",
 		   disabled_links, link_mapping_size, switch_time, duration);
-- 
2.45.2

